### PR TITLE
Remove exported `getReport()`

### PR DESCRIPTION
**Pull request by gijoe0295**
### Details

Remove exported `getReport`.

### Fixed Issues

$ https://github.com/Expensify/App/issues/40316
PROPOSAL: https://github.com/Expensify/App/issues/40316#issuecomment-2059982364

### Tests

1. Test basic features of reports and report actions: New report, Order of reports in LHN, Send/Delete/Edit/Reply/Mark as unread message, Message context menu, Message notifications, Quick action, Hold IOU request,...
2. Some edge cases: https://github.com/Expensify/App/pull/38884, https://github.com/Expensify/App/pull/34320, https://github.com/Expensify/App/pull/28702/files#r1343844837

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

### QA Steps

1. Test basic features of reports and report actions: New report, Order of reports in LHN, Send/Delete/Edit/Reply/Mark as unread message, Message context menu, Message notifications, Quick action, Hold IOU request,...

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/153004152/9df6a593-9f6f-4d4e-8d24-2c0b2873bbfc

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot_1713729984

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/153004152/19e9e4e8-f1ba-4f6d-8b37-4c00ae2a068a

https://github.com/Expensify/App/assets/153004152/01266a48-f562-4132-93ce-c1c1aa752e61

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/153004152/200982fc-3c24-4a63-93ff-6e3222ec8657

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-04-22 at 02 26 07
Screenshot 2024-04-22 at 02 39 43
Screenshot 2024-04-22 at 02 40 42
Screenshot 2024-04-22 at 02 43 41
Screenshot 2024-04-22 at 02 48 05
Screenshot 2024-04-22 at 02 49 08
Screenshot 2024-04-22 at 02 49 44

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/153004152/8015b3d7-7b43-4c72-a19a-3a4877b4627b

</details>

### CODE REVIEW

File: src/libs/ReportUtils.ts [hunk 1]
```diff
@@ -1144,7 +1145,8 @@ function isClosedExpenseReportWithNoExpenses(report: OnyxEntry<Report>): boolean
 /**
  * Whether the provided report is an archived room
  */
-function isArchivedRoom(report: OnyxEntry<Report> | EmptyObject): boolean {
+function isArchivedRoom(reportOrID: OnyxEntry<Report> | EmptyObject | string): boolean {
+    const report = typeof reportOrID === 'string' ? allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${reportOrID}`] ?? null : reportOrID;
```
Comment (s77rt) [inline src/libs/ReportUtils.ts:1149]: Revert. This is same as getReport + isArchivedRoom

**Review by s77rt — CHANGES_REQUESTED** [review]
+ Resolve conflicts

File: src/pages/home/ReportScreen.tsx [hunk 2]
```diff
@@ -717,6 +722,14 @@ function ReportScreen({
...
 ReportScreen.displayName = 'ReportScreen';
 
+const parentReportSelector = (report: OnyxEntry<OnyxTypes.Report>): OnyxEntry<OnyxTypes.Report> =>
+    report && {
+        type: report.type,
+        reportID: report.reportID,
+        parentReportID: report.parentReportID,
+        parentReportActionID: report.parentReportActionID,
+    };
+
```
[the remark this answers is not in the document]
Comment (s77rt) [reply src/pages/home/ReportScreen.tsx]: That sounds good to me

File: src/pages/home/report/ReportActionItem.tsx [hunk 3]
```diff
@@ -196,11 +201,25 @@ function ReportActionItem({
...
     const transactionCurrency = TransactionUtils.getCurrency(transaction);
     const reportScrollManager = useReportScrollManager();
 
+    // A performance improvement. Put this in withOnyx would increase rerender count massively.
+    const [originalReport, setOriginalReport] = useState<OnyxEntry<OnyxTypes.Report>>(report);
+    useEffect(() => {
+        if (!originalReportID || originalReportID === report.reportID) {
+            return;
+        }
+        const unsubscribeOnyx = onyxSubscribe({
```
[the remark this answers is not in the document]
Comment (s77rt) [reply src/pages/home/report/ReportActionItem.tsx:214]: `Onyx.connect` should not be used in components. Can you check for other options?

[the remark this answers is not in the document]
Comment (s77rt) [reply src/pages/home/report/ReportActionItem.tsx:214]: The usage there also does not follow our Onyx practices.

**Comment by s77rt** [comment]
@gijoe0295 Any update on the use of `onyxSubscribe`?

File: src/pages/home/report/ReportActionItem.tsx [hunk 4]
```diff
@@ -555,7 +574,7 @@ function ReportActionItem({
                 </ShowContextMenuContext.Provider>
             );
         } else if (action.actionName === CONST.REPORT.ACTIONS.TYPE.REIMBURSEMENT_QUEUED) {
-            const linkedReport = ReportUtils.isChatThread(report) ? ReportUtils.getReport(report.parentReportID) : report;
+            const linkedReport = ReportUtils.isChatThread(report) ? parentReport : report;
```
Comment (s77rt) [inline src/pages/home/report/ReportActionItem.tsx:577]: Looks we can get ride of the `parentReport`. We only need 2 props from the parent
- reportID which can be replaced by report.parentReportID
- ownerAccountID write an util function for that

**Review by s77rt — CHANGES_REQUESTED** [review]

**Comment by s77rt** [comment]
@gijoe0295 Any updates on this?

**Comment by tgolen** [comment]
@gijoe0295 What is the status of this PR? Can you please finish it up?

**Comment by s77rt** [comment]
@gijoe0295 Any updates here?

**Comment by tgolen** [comment]
@gijoe0295 What can I do to help get this moving along?

**Pull request closed without merging**